 # Building user-based recommendation model for Amazon

Hello to you, <br>
Sanjay Kumar here, I'm about to start my first ML project. So I recommend the team to correct my mistakes without fail and in case if you are satisfied with my work, then you can Suggest me more ideas to get my results even Better. 

### Import  Basic Librabries  

In [89]:
import numpy as np
import pandas as pd 

### Get  Data 

In [90]:
df = pd.read_csv(r'K:\SK\Simplilearn\Machine Learning\ML projects\Movie recommendation project\1567507480_amazonmoviesandtvratings\Amazon - Movies and TV Ratings.csv')

In [129]:
df.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     NaN     NaN     NaN     NaN     NaN   
1   AH3QC2PC1VTGP     NaN     NaN     2.0     NaN     NaN     NaN     NaN   
2  A3LKP6WPMP9UKX     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
3   AVIY68KEPQ5ZD     NaN     NaN     NaN     5.0     NaN     NaN     NaN   
4  A1CV1WROP5KTTW     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
1     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
2     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
3     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 207 columns]

In [130]:
df.shape

(4848, 207)

In [131]:
df.dtypes

user_id      object
Movie1      float64
Movie2      float64
Movie3      float64
Movie4      float64
             ...   
Movie202    float64
Movie203    float64
Movie204    float64
Movie205    float64
Movie206    float64
Length: 207, dtype: object

### EDA 

First Let's Conevert NaN Values to 0, As rating vary from 1-10, Let's consider 0 has user either Not watched or Not rated the movie. 

In [132]:
df1 = df.fillna(0)

In [133]:
df1.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     0.0     0.0     0.0     0.0     0.0   
1   AH3QC2PC1VTGP     0.0     0.0     2.0     0.0     0.0     0.0     0.0   
2  A3LKP6WPMP9UKX     0.0     0.0     0.0     5.0     0.0     0.0     0.0   
3   AVIY68KEPQ5ZD     0.0     0.0     0.0     5.0     0.0     0.0     0.0   
4  A1CV1WROP5KTTW     0.0     0.0     0.0     0.0     5.0     0.0     0.0   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
1     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
2     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
3     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
4     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       0.0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0       0.0  

[5 rows x 207 columns]

Now we have converted all the values of NaN to Zero. <BR>
Since user id is unique value for all we can remove this coloumn

In [134]:
df1 = df1.drop(['user_id'], axis= 1)
df1.head()

Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  Movie8  Movie9  \
0     5.0     5.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1     0.0     0.0     2.0     0.0     0.0     0.0     0.0     0.0     0.0   
2     0.0     0.0     0.0     5.0     0.0     0.0     0.0     0.0     0.0   
3     0.0     0.0     0.0     5.0     0.0     0.0     0.0     0.0     0.0   
4     0.0     0.0     0.0     0.0     5.0     0.0     0.0     0.0     0.0   

   Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  Movie202  \
0      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
4      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   Movie203  Movie204  Movie205  Movie206  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 206 columns]

Let's Convert Data type from float to int 


In [135]:
df1 = df1.astype(int)
df1.dtypes

Movie1      int32
Movie2      int32
Movie3      int32
Movie4      int32
Movie5      int32
            ...  
Movie202    int32
Movie203    int32
Movie204    int32
Movie205    int32
Movie206    int32
Length: 206, dtype: object

In [136]:
df1.head()

Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  Movie8  Movie9  \
0       5       5       0       0       0       0       0       0       0   
1       0       0       2       0       0       0       0       0       0   
2       0       0       0       5       0       0       0       0       0   
3       0       0       0       5       0       0       0       0       0   
4       0       0       0       0       5       0       0       0       0   

   Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  Movie202  \
0        0  ...         0         0         0         0         0         0   
1        0  ...         0         0         0         0         0         0   
2        0  ...         0         0         0         0         0         0   
3        0  ...         0         0         0         0         0         0   
4        0  ...         0         0         0         0         0         0   

   Movie203  Movie204  Movie205  Movie206  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 206 columns]

# Which movies have maximum views/ratings?

In [137]:
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('white')
%matplotlib inline

In [152]:
res = []
for col in df1.columns:
    L1 = df1[col].tolist()
    res.append(L1)
    

    
   

In [141]:
movie1list = res[15]


In [142]:
i=0 
for x in movie1list:
    if x!=0:
        i=i+1
    else:
        pass
print(i)

320


In [143]:
Movielist =[]
for x in res:
    i=0
    for y in x:
        if y!=0:
            i=i+1
        else:
            pass
    Movielist.append(i)
print(Movielist)

[1, 1, 1, 2, 29, 1, 1, 1, 1, 1, 2, 5, 1, 1, 1, 320, 1, 1, 2, 1, 1, 2, 3, 5, 1, 3, 1, 3, 243, 2, 2, 2, 1, 1, 1, 1, 1, 1, 4, 3, 1, 1, 7, 2, 1, 1, 1, 1, 1, 1, 2, 17, 3, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 2, 1, 12, 2, 1, 1, 3, 21, 1, 1, 83, 18, 128, 101, 2, 3, 6, 3, 5, 1, 2, 1, 5, 2, 272, 4, 3, 1, 39, 54, 13, 8, 39, 2, 4, 7, 1, 1, 11, 5, 8, 3, 4, 4, 1, 4, 5, 2, 2313, 3, 3, 4, 4, 3, 1, 6, 1, 2, 3, 13, 4, 578, 7, 1, 1, 1, 1, 1, 1, 2, 1, 3, 2, 1, 1, 1, 2, 1, 4, 66, 2, 6, 25, 15, 13, 4, 1, 2, 4, 4, 4, 3, 1, 2, 15, 4, 1, 1, 1, 1, 7, 1, 2, 30, 1, 17, 24, 9, 1, 6, 5, 7, 6, 10, 7, 7, 1, 9, 5, 2, 1, 8, 3, 6, 1, 8, 35, 13]


In [144]:
Max_viewer = max(Movielist)

Max_index = Movielist.index(Max_viewer)
print(Max_index)

126


In [145]:
S = 0
value =[] 
for x in Movielist:
    if x < 240:
        pass
    else:
        S = Movielist.index(x)
    value.append(S)
value = list(dict.fromkeys(value))
print(value)


[0, 15, 28, 102, 126, 139]


## With 2313 Views "Movie_127"  has the maximum Views/ Ratings. 

In [107]:
Movielist.sort()
print(Movielist)


[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 9, 9, 10, 11, 12, 13, 13, 13, 13, 15, 15, 17, 17, 18, 21, 24, 25, 29, 30, 35, 39, 39, 54, 66, 83, 101, 128, 243, 272, 320, 578, 2313]


In [108]:
max(Movielist)

2313

In [109]:
High_Movie_views = print(Movielist[-5:])

[243, 272, 320, 578, 2313]


#### Movie_16, Movie_29, Movie_103, movie_127 and movie_140 are top five highest viewed/ rated Movies 

### What is the average rating for each movie? Define the top 5 movies with the maximum ratings.

In [110]:
max(Movielist)

2313

In [150]:
Movielist =[]
for x in res:
    total = len(x)
    S = sum(x)
    res = S/total
 

In [153]:
Avg_movie_list =[]
for x in res:
    i=0
    for y in x:
        if y!=0:
            i=i+1
        else:
            pass
#     Movielist.append(i)
    total = i 
    s = sum(x)
    if total ==0:
        pass
    else:
        res = s/total
        Avg_movie_list.append(res)
print(Avg_movie_list)



[5.0, 5.0, 2.0, 5.0, 4.103448275862069, 4.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.0, 5.0, 4.51875, 3.0, 5.0, 3.5, 3.0, 5.0, 5.0, 5.0, 4.4, 5.0, 3.0, 5.0, 3.3333333333333335, 4.806584362139918, 4.5, 5.0, 4.5, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.857142857142857, 5.0, 1.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.5, 3.4705882352941178, 2.0, 5.0, 5.0, 5.0, 5.0, 1.0, 2.0, 1.0, 5.0, 3.0, 5.0, 3.0, 5.0, 5.0, 1.0, 5.0, 1.0, 5.0, 4.0, 5.0, 2.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.0, 4.416666666666667, 5.0, 3.0, 5.0, 5.0, 4.666666666666667, 5.0, 4.0, 4.578313253012048, 1.8333333333333333, 4.578125, 4.772277227722772, 5.0, 4.333333333333333, 3.3333333333333335, 5.0, 4.8, 5.0, 4.0, 4.0, 5.0, 4.0, 4.5625, 4.5, 5.0, 5.0, 4.0, 4.666666666666667, 4.384615384615385, 4.875, 4.641025641025641, 5.0, 3.75, 4.428571428571429, 4.0, 5.0, 4.7272727272727275, 5.0, 4.375, 5.0, 4.25, 5.0, 5.0, 4.75, 4.8, 4.5, 4.11197578901859, 5.0, 4.0, 4.5, 5.0, 5.0, 5.0, 4.0, 5.0, 5.0, 4.0, 4.076923076923077, 5.0, 4.8339100346020

In [154]:
len(Avg_movie_list)

206

### Average rating for each movie is shown in the list above, Movies with only one user rating is ignored

In [117]:
R=0
High_movie_list =[]
for x in res:
    i=0
    for y in x:
        if y!=0:
            i=i+1
        else:
            pass
#     Movielist.append(i)
    total = i 
    s = sum(x)
    if total < 50:
        pass
    
    else:
        res = s/total
        High_movie_list.append(res)
print(High_movie_list)

[4.51875, 4.806584362139918, 4.578313253012048, 4.578125, 4.772277227722772, 4.5625, 4.666666666666667, 4.11197578901859, 4.833910034602076, 4.818181818181818]


In [118]:
len(High_movie_list)

10

In [119]:
High_movie_list = [round(num, 3) for num in High_movie_list]
print(High_movie_list)

[4.519, 4.807, 4.578, 4.578, 4.772, 4.562, 4.667, 4.112, 4.834, 4.818]


In [120]:
S = 0
R= 0
value =[] 
Rating = []
for x in Movielist:
    if x < 50:
        pass
    else:
        S = Movielist.index(x)
    value.append(S)
    Rating.append(R)
    
value = list(dict.fromkeys(value))
print(value)




[]


In [121]:
len(value)

0

In [122]:
value = del value[0]

SyntaxError: invalid syntax (<ipython-input-122-a211f956252a>, line 1)

In [48]:
print(value)

[]


In [123]:
print(value)
print(High_movie_list)
ser = {}
for key in value:
    for rating in High_movie_list:
        ser[key] = rating
        High_movie_list.remove(rating)
        break
    print(ser)


[]
[4.519, 4.807, 4.578, 4.578, 4.772, 4.562, 4.667, 4.112, 4.834, 4.818]


# Recommendation Model: Some of the movies hadn’t been watched and therefore, are not rated by the users. Netflix would like to take this as an opportunity and build a machine learning recommendation algorithm which provides the ratings for each of the users

## Divide the data into training and test data

In [63]:
df.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     0.0     0.0     0.0     0.0     0.0   
1   AH3QC2PC1VTGP     0.0     0.0     2.0     0.0     0.0     0.0     0.0   
2  A3LKP6WPMP9UKX     0.0     0.0     0.0     5.0     0.0     0.0     0.0   
3   AVIY68KEPQ5ZD     0.0     0.0     0.0     5.0     0.0     0.0     0.0   
4  A1CV1WROP5KTTW     0.0     0.0     0.0     0.0     5.0     0.0     0.0   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
1     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
2     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
3     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
4     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       0.0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0       0.0  

[5 rows x 207 columns]

In [64]:
df = df.fillna(0)

In [65]:
df.head()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
0  A3R5OBKS7OM2IR     5.0     5.0     0.0     0.0     0.0     0.0     0.0   
1   AH3QC2PC1VTGP     0.0     0.0     2.0     0.0     0.0     0.0     0.0   
2  A3LKP6WPMP9UKX     0.0     0.0     0.0     5.0     0.0     0.0     0.0   
3   AVIY68KEPQ5ZD     0.0     0.0     0.0     5.0     0.0     0.0     0.0   
4  A1CV1WROP5KTTW     0.0     0.0     0.0     0.0     5.0     0.0     0.0   

   Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
0     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
1     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
2     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
3     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   
4     0.0     0.0  ...       0.0       0.0       0.0       0.0       0.0   

   Movie202  Movie203  Movie204  Movie205  Movie206  
0       0.0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0       0.0  

[5 rows x 207 columns]

In [66]:
import matplotlib.pyplot as mp
import pandas as pd
import seaborn as sb

In [80]:
Movie16_ratings = df['Movie16']
Movie29_ratings = df['Movie29']
Movie140_ratings = df['Movie140']
Movie139_ratings = df['Movie139']

In [81]:
similar_to_Movie139 = df.corrwith(Movie139_ratings)
similar_to_Movie140 = df.corrwith(Movie140_ratings)
similar_to_Movie16 = df.corrwith(Movie16_ratings)
similar_to_Movie29 = df.corrwith(Movie29_ratings)

Let's clean this by removing NaN values and using a DataFrame instead of a series:

In [82]:
corr_Movie139 = pd.DataFrame(similar_to_Movie139,columns=['Correlation'])
corr_Movie140 = pd.DataFrame(similar_to_Movie140,columns=['Correlation'])
corr_Movie16 = pd.DataFrame(similar_to_Movie16,columns=['Correlation'])
corr_Movie29 = pd.DataFrame(similar_to_Movie29, columns=['Correlation'])

In [83]:
corr_Movie140.sort_values('Correlation', ascending = False, inplace = True)
corr_Movie139.sort_values('Correlation', ascending = False, inplace = True)
corr_Movie16.sort_values('Correlation', ascending = False, inplace = True)
corr_Movie29.sort_values('Correlation', ascending = False, inplace = True)


In [84]:
corr_Movie139.drop('Movie139' , inplace = True)
corr_Movie140.drop('Movie140' , inplace = True)
corr_Movie16.drop('Movie16' , inplace = True)
corr_Movie29.drop('Movie29' , inplace = True)


In [85]:
corr_Movie139 = corr_Movie139[corr_Movie139['Correlation']>0]
print('Movies Recommended if the use watch Movie139 are as follows')
print(corr_Movie139)

Movies Recommended if the use watch Movie139 are as follows
          Correlation
Movie138     0.160558
Movie163     0.151509


In [86]:
corr_Movie140 = corr_Movie140[corr_Movie140['Correlation']>0]
print('Movies Recommended if the use watch Movie140 are as follows')
print(corr_Movie140)

Movies Recommended if the use watch Movie140 are as follows
          Correlation
Movie189     0.028944
Movie198     0.024727
Movie44      0.024727
Movie188     0.024285
Movie191     0.024285
Movie190     0.022343


In [87]:
corr_Movie16 = corr_Movie16[corr_Movie16['Correlation']>0]
print('Movies Recommended if the use watch Movie16 are as follows')
print(corr_Movie16)

Movies Recommended if the use watch Movie16 are as follows
          Correlation
Movie162     0.002148


In [88]:
corr_Movie29 = corr_Movie29[corr_Movie29['Correlation']>0]
print('Movies Recommended if the use watch Movie29 are as follows')
print(corr_Movie29)

Movies Recommended if the use watch Movie29 are as follows
Empty DataFrame
Columns: [Correlation]
Index: []


# Summarising insights.


Model-based Collaborative Filtering is a personalised recommender system, the recommendations are based on the past behavior of the user and it is not dependent on any additional information.